pkg/gofr/service/ratelimiter.go

```diff
@@ -0,0 +1,475 @@
+
+		return context.WithCancel(newCtx)
+	}
+
+	return context.WithCancel(ctx)
+}
+
+// enqueueRequest adds a request to the rate limiter queue and waits for its response.
+func (r *RateLimiter) enqueueRequest(ctx context.Context, execute func() (*http.Response, error)) (*http.Response, error) {
+	ctx, cancel := handleContextDeadline(ctx)
+	defer cancel()
+
+	span := trace.SpanFromContext(ctx)
+	traceID := span.SpanContext().TraceID().String()
+	startTime := time.Now()
+
+	r.mutex.Lock()
+	if r.isShuttingDown {
+		r.mutex.Unlock()
+		r.Log(&ErrorLog{
+			Log: &Log{
+				Timestamp: time.Now(), CorrelationID: traceID, ResponseTime: time.Since(startTime).Microseconds(),
+				ResponseCode: http.StatusServiceUnavailable,
+				HTTPMethod:   "REQUEST",
+				URI:          "rate_limiter",
+			},
+			ErrorMessage: "rate limiter shutting down",
+		})
+
+		return nil, fmt.Errorf("rate limiter shutting down for request with TraceID %s: %w", traceID, errShuttingDown)
+	}
+
+	currentQueueSize := atomic.LoadInt32(&r.currentQueue)
+	if currentQueueSize >= int32(r.maxQueue) {
+		r.mutex.Unlock()
+		r.Log(&ErrorLog{
+			Log: &Log{
+				Timestamp: time.Now(), CorrelationID: traceID, ResponseTime: time.Since(startTime).Microseconds(),
+				ResponseCode: http.StatusServiceUnavailable,
+				HTTPMethod:   "REQUEST",
+				URI:          "rate_limiter",
+			},
+			ErrorMessage: fmt.Sprintf("queue full (max size: %d, current size: %d)", r.maxQueue, currentQueueSize),
+		})
+
+		return nil, fmt.Errorf("%w for request with TraceID %s (max size: %d, current size: %d)",
+			errQueueFull, traceID, r.maxQueue, currentQueueSize)
+	}
+	r.mutex.Unlock()
+
+	respCh := make(chan *requestResponse, 1)
+	reqCtx, reqCancel := context.WithCancel(ctx)
+
+	if !atomic.CompareAndSwapInt32(&r.currentQueue, currentQueueSize, currentQueueSize+1) {
+		reqCancel()
+		return nil, fmt.Errorf("%w for request with TraceID %s (queue size changed)", errQueueFull, traceID)
+	}
+
+	r.wg.Add(1)
+
+	req := requestWrapper{
+		execute: execute,
+		respCh:  respCh,
+		ctx:     reqCtx,
+		cancel:  reqCancel,
+	}
+
+	select {
+	case r.requests <- req:
+		r.Log(&Log{
+			Timestamp: time.Now(), CorrelationID: traceID, ResponseTime: time.Since(startTime).Microseconds(),
+			ResponseCode: http.StatusAccepted,
+			HTTPMethod:   "REQUEST",
+			URI:          "rate_limiter",
+		})
+
+		return r.handleResponse(ctx, respCh, reqCancel)
+
+	case <-ctx.Done():
+		atomic.AddInt32(&r.currentQueue, -1)
+		r.wg.Done()
+		reqCancel()
+		r.Log(&ErrorLog{
+			Log: &Log{Timestamp: time.Now(), CorrelationID: traceID, ResponseTime: time.Since(startTime).Microseconds(),
+				ResponseCode: http.StatusGatewayTimeout,
+				HTTPMethod:   "REQUEST",
+				URI:          "rate_limiter",
+			},
+			ErrorMessage: fmt.Sprintf("context canceled while enqueueing: %v", ctx.Err()),
+		})
+
+		return nil, fmt.Errorf("context canceled while enqueueing with TraceID %s: %w", traceID, ctx.Err())
+
+	case <-r.done:
+		atomic.AddInt32(&r.currentQueue, -1)
```

Comment:
This usage and syntax of sync/atomic is almost deprecated because it is too complicated to use.

Take a look at atomic.Int32 for currentQueue instead of int32 and replace all the code accordingly